File: src/main/java/com/projectswg/forwarder/resources/networking/data/Packager.java
package com.projectswg.forwarder.resources.networking.data;

import com.projectswg.forwarder.resources.networking.data.ProtocolStack.ConnectionStream;
import com.projectswg.forwarder.resources.networking.packets.DataChannel;
import com.projectswg.forwarder.resources.networking.packets.Fragmented;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.atomic.AtomicInteger;

public class Packager {
	
	private final AtomicInteger size;
	private final List<byte[]> dataChannel;
	private final BlockingQueue<byte[]> outboundRaw;
	private final ConnectionStream<SequencedOutbound> outboundPackaged;
	
	public Packager(BlockingQueue<byte[]> outboundRaw, ConnectionStream<SequencedOutbound> outboundPackaged, ProtocolStack stack) {
		this.size = new AtomicInteger(8);
		this.dataChannel = new ArrayList<>();
		this.outboundRaw = outboundRaw;
		this.outboundPackaged = outboundPackaged;
	}
	
	public void handle(int maxPackaged) {
		byte [] packet;
		int packetSize;
		
		while (outboundPackaged.size() < maxPackaged) {
			packet = outboundRaw.poll();
			if (packet == null)
				break;
			
			packetSize = getPacketLength(packet);

			if (size.get() + packetSize >= 16384) // max data channel size
				sendDataChannel();

			if (packetSize < 16384) { // if overflowed, must go into fragmented
				addToDataChannel(packet, packetSize);
			} else {
				sendFragmented(packet);
			}
		}
		sendDataChannel();
	}
	
	private void addToDataChannel(byte [] packet, int packetSize) {
		dataChannel.add(packet);
		size.getAndAdd(packetSize);
	}
	
	private void sendDataChannel() {
		if (dataChannel.isEmpty())
			return;
		
		outboundPackaged.addOrdered(new SequencedOutbound(new DataChannel(dataChannel)));
		reset();
	}
	
	private void sendFragmented(byte [] packet) {
		byte[][] frags = Fragmented.split(packet);
		for (byte [] frag : frags) {
			outboundPackaged.addOrdered(new SequencedOutbound(new Fragmented((short) 0, frag)));
		}
	}
	
	private void reset() {
		dataChannel.clear();
		size.set(8);
	}
	
	private static int getPacketLength(byte [] data) {
		int len = data.length;
		if (len >= 255)
			return len + 3;
		return len + 1;
	}
	
}

File: src/main/java/com/projectswg/forwarder/resources/networking/data/Packager.kt
package com.projectswg.forwarder.resources.networking.data

import com.projectswg.forwarder.resources.networking.data.ProtocolStack.ConnectionStream
import com.projectswg.forwarder.resources.networking.packets.DataChannel
import com.projectswg.forwarder.resources.networking.packets.Fragmented
import java.util.*
import java.util.concurrent.BlockingQueue
import java.util.concurrent.atomic.AtomicInteger

class Packager(private val outboundRaw: BlockingQueue<ByteArray>, private val outboundPackaged: ConnectionStream<SequencedOutbound>) {
	
	private val size: AtomicInteger = AtomicInteger(8)
	private val dataChannel: MutableList<ByteArray> = ArrayList()
	
	fun handle(maxPackaged: Int) {
		var packet: ByteArray?
		var packetSize: Int
		
		while (outboundPackaged.size() < maxPackaged) {
			packet = outboundRaw.poll()
			if (packet == null)
				break
			
			packetSize = getPacketLength(packet)
			
			if (size.get() + packetSize >= 16384) // max data channel size
				sendDataChannel()
			
			if (packetSize < 16384) { // if overflowed, must go into fragmented
				addToDataChannel(packet, packetSize)
			} else {
				sendFragmented(packet)
			}
		}
		sendDataChannel()
	}
	
	private fun addToDataChannel(packet: ByteArray, packetSize: Int) {
		dataChannel.add(packet)
		size.getAndAdd(packetSize)
	}
	
	private fun sendDataChannel() {
		if (dataChannel.isEmpty())
			return
		
		outboundPackaged.addOrdered(SequencedOutbound(DataChannel(dataChannel)))
		reset()
	}
	
	private fun sendFragmented(packet: ByteArray) {
		val frags = Fragmented.split(packet)
		for (frag in frags) {
			outboundPackaged.addOrdered(SequencedOutbound(Fragmented(0.toShort(), frag)))
		}
	}
	
	private fun reset() {
		dataChannel.clear()
		size.set(8)
	}
	
	private fun getPacketLength(data: ByteArray): Int {
		val len = data.size
		return if (len >= 255) len + 3 else len + 1
	}
	
}
